python/lsst/meas/astrom/matchOptimisticB.py

```diff
@@ -0,0 +1,241 @@
+        min = 2,
+    )
+    minMatchedPairs = pexConfig.RangeField(
+        doc = "Minimum number of matched pairs; see also minFracMatchedPairs",
+        dtype = int,
+        default = 30,
+        min = 2,
+    )
+    minFracMatchedPairs = pexConfig.RangeField(
+        doc = "Minimum number of matched pairs as a fraction of the smaller of "
+            "the number of reference stars or the number of good sources; "
+            "the actual minimum is the smaller of this value or minMatchedPairs",
+        dtype = float,
+        default = 0.3,
+        min = 0,
+        max = 1,
+    )
+    maxOffsetPix = pexConfig.RangeField(
+        doc = "Maximum offset between sources and position reference objects (pixel)",
+        dtype = int,
+        default = 300,
+        max = 4000,
+    )
+    rotationAllowedInRad = pexConfig.RangeField(
+        doc = "Rotation angle allowed between sources and position reference objects (radian)",
+        dtype = float,
+        default = 0.02,
+        max = 0.1,
+    )
+    angleDiffFrom90 = pexConfig.RangeField(
+        doc = "Difference of angle between x and y from 90 degree allowed (degree)",
+        dtype = float,
+        default = 0.2,
+        max = 45.0,
+    )
+    numPointsForShape = pexConfig.Field(
+        doc = "number of points to define a shape for matching",
+        dtype = int,
+        default = 6,
+    )
+    maxDeterminant = pexConfig.Field(
+        doc = "maximum determinant of linear transformation matrix for a usable solution",
+        dtype = float,
+        default = 0.02,
+    )
+
+def isCleanSource(s):
+    """!Return True if star has known centroid and not at the edge
+    """
+    return (numpy.isfinite(s.getX()) and numpy.isfinite(s.getY()) and not s.get("base_PixelFlags_flag_edge"))
+
+def isGoodSource(s):
+    """!Return True if star is "clean" (see isCleanSource) and not saturated
+    """
+    # FIXME: should use Key to get flag (but then we'd need schema in advance)
+    return (isCleanSource(s) and not s.getCentroidFlag() and not s.get("base_PixelFlags_flag_saturated"))
+
+
+class MatchOptimisticBTask(pipeBase.Task):
+    """!Match sources to position reference objects using optimistic pattern matching
+
+    See the matchOptimisticB.h for information about the matching algorithm.
+    """
+    ConfigClass = MatchOptimisticBConfig
+    _DefaultName = "matchObjectsToSources"
+
+    @staticmethod
+    def _purgeM31M32Stars(starCat):
+        """!Remove stars from a catalog that are too near the center of M31 or M32
+        """
+        retStarCat = type(starCat)(starCat.table)
+        m31 = afwCoord.Coord(afwGeom.Point2D(10.681, 41.269))
+        m32 = afwCoord.Coord(afwGeom.Point2D(10.675, 40.864))
+        for star in starCat:
+            if (star.getCoord().angularSeparation(m31).asArcminutes() > 2.2 and
+                star.getCoord().angularSeparation(m32).asArcminutes() > 1.0):
+                retStarCat.append(star)
+        return retStarCat
```

Comment:
I suspect this code is carried over from some special-case processing of HSC M31 data, and I don't think it really belongs in a general-purpose matcher.  Or, rather, since it's clear we need to do this kind of rejection in some fields, it'd be better to make it somehow configurable rather than based on hard-coded positions.